### PR TITLE
[Federation] Convert the ReplicaSet controller to a sync controller.

### PR DESCRIPTION
**Pull request by perotinus**
See #45563 for previous discussion: that PR was split into two, with this one containing the actual conversion work and that one containing the implementation of the scheduling methods in the sync controller.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by perotinus** [comment]
@k8s-bot test this

**Comment by perotinus** [comment]
The CLA should already be signed.

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 pull-kubernetes-federation-e2e-gce test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Review by marun — COMMENTED** [review]
Hmm, neglected to publish this previously...

Would be good to see #45563 merge before reviewing in-depth.  My quibbles about testing aside, though, I think this is ok so long as the integration and e2e tests pass.

File: federation/pkg/federation-controller/sync/replicasetcontroller_test.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package replicaset
+package sync
```
Comment (perotinus) [inline federation/pkg/federation-controller/sync/replicasetcontroller_test.go]: I'm not quite sure. I think it is testing a bit more than the current integration tests–namely, that the scheduling causes the replicas to be distributed to clusters such that the numbers add up–but it's certainly a test that would be better removed in the future and replaced with unit tests for scheduling and integration tests for propagation.

File: federation/pkg/federatedtypes/replicaset.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+		}
+		_, err := a.client.Extensions().ReplicaSets(rs.Namespace).UpdateStatus(rs)
+		return err
+	}
+	return nil
+}
+
+func (a *ReplicaSetAdapter) schedule(frs *extensionsv1.ReplicaSet, clusterNames []string,
+	current map[string]int64, estimatedCapacity map[string]int64) map[string]int64 {
+	// TODO: integrate real scheduler
```
Comment (perotinus) [inline federation/pkg/federatedtypes/replicaset.go]: Yep, this is a TODO from the previous code.

File: federation/pkg/federatedtypes/replicaset.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+		glog.Info("Invalid ReplicaSet specific preference, use default. rs: %v, err: %v", frs, err)
+	}
+	if frsPref != nil { // create a new planner if user specified a preference
+		plnr = planner.NewPlanner(frsPref)
+	}
+
+	replicas := int64(*frs.Spec.Replicas)
+	scheduleResult, overflow := plnr.Plan(replicas, clusterNames, current, estimatedCapacity,
+		frs.Namespace+"/"+frs.Name)
+	// make sure the return contains clusters need to zero the replicas
```
Comment (perotinus) [inline federation/pkg/federatedtypes/replicaset.go]: Done.

File: federation/pkg/federatedtypes/replicaset_test.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		pod1Phase                     apiv1.PodPhase
+		pod1Condition                 apiv1.PodCondition
+		pod2Phase                     apiv1.PodPhase
+		pod2Condition                 apiv1.PodCondition
+		cluster1Replicas              *int64
+		cluster2Replicas              *int64
+		cluster1UnschedulableReplicas *int64
+		cluster2UnschedulableReplicas *int64
+	}{
+		{rs1Replicas: 2, rs2Replicas: 2, rs1ReadyReplicas: 2, rs2ReadyReplicas: 2, podsGetter: uncalledPodsGetter, cluster1Replicas: &two, cluster2Replicas: &two},
```
Comment (perotinus) [inline federation/pkg/federatedtypes/replicaset_test.go]: Done.

File: federation/pkg/federatedtypes/replicaset_test.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+const (
+	pods        = "pods"
+	replicasets = "replicasets"
+	k8s1        = "k8s-1"
+	k8s2        = "k8s-2"
+)
+
+func TestClusterReplicaState(t *testing.T) {
```
Comment (perotinus) [inline federation/pkg/federatedtypes/replicaset_test.go]: Hmm. I thought about it for a little while, but I'm concerned that I couldn't think of a way to do so without making this more convoluted. I'll think about this a bit more, perhaps in the context of the deployments PR that follows this one up.

File: federation/pkg/federation-controller/sync/replicasetcontroller_test.go (hunk 1 above)
Comment (marun) [inline federation/pkg/federation-controller/sync/replicasetcontroller_test.go]: I'd hope the numbers adding up could be validated separately from propagation, but sure.

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by marun** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by csbell** [comment]
Assuming remaining issues will be resolved and Maru's lgtm
/approve

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by fejta** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref: #46827

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by perotinus** [comment]
@marun Can you LGTM this?

**Comment by perotinus** [comment]
Related issue: #39986

**Comment by perotinus** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by perotinus** [comment]
Federation E2E tests passed: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/46527/pull-kubernetes-federation-e2e-gce/9670/

**Pull request merged**